reworked key combiner to be NIST conforming

falko-strenzke opened the pull request:
This PR brings the following changes:

- Revert the key combiner KDF back to KMAC256. Reason: This is currently the only known way to combine two keys in a NIST-compliant way without modifying the ECC-KEM. See #132 and [issue 10 of the NIST/Brainpool draft](https://github.com/openpgp-pqc/draft-ehlen-openpgp-nist-bp-comp/issues/10) for details.
- This change also implictly fixes the input-order-problem of the key combiner raised in #127.
- It also implicitly fixes a potential problem regarding the KDF input string being uniquely parsable from the rear raised in #127.
- We remove the counter from the KDF. This is justified by
  - https://csrc.nist.gov/csrc/media/Projects/cryptographic-module-validation-program/documents/fips%20140-3/FIPS%20140-3%20IG.pdf#page=178 which explicitly addresses the general counter removal for KDFs that use only a single iteration
  - and specificly for KMAC as a KDF in [NIST.SP.800-108r1-upd1](https://nvlpubs.nist.gov/nistpubs/SpecialPublications/NIST.SP.800-108r1-upd1.pdf#page=19).
- We remove the `fixedInfo` field and directly replace it with `algID`. This is merely an editorial change.

Fixes #132.
Fixes #127.

Once this is approved we will apply the same construction in https://github.com/openpgp-pqc/draft-ehlen-openpgp-nist-bp-comp.

In the meeting on 2024-08-22, we agreed to create new test vectors only before the publication.

# Open work items

Before we can remove the draft status, there is one thing that needs checking regarding the test vectors:
- [x] @wussler @TJ-91  Please check the naming of the key share intermediate values. There was a label "SHA3-256" that I couldn't make any sense of and which I removed. I also added "input to `multiKeyCombine`" as I assume this is was is meant here. There are still "(??)" marks in the text for this reason which we should remove before merging when my changes have been verified by you.

TJ-91 commented on draft-ietf-openpgp-pqc.md:
```diff
@@ -972,10 +946,10 @@ In the random oracle setting, the reordering does not influence the arguments in
 This ensures that the input keying material is used to generate a KEK for a specific purpose or context.
 
-The `algID` defined in {{kem-fixed-info}} binds the derived KEK to the chosen algorithm and communication parties.
+The `algID` defined in {{kem-key-combiner}} binds the derived KEK to the chosen algorithm and communication parties.
```
I think this wording was more natural when it referred to `fixedInfo`. Now it's weird to refer to the `algID` input like this so we should rephrase.

wussler replied:
```suggestion
The `algID` defined in {{kem-key-combiner}} binds the derived KEK to the chosen algorithm.
```

falko-strenzke replied:
I had already made the above suggestion, only that I was still "review mode" for some reason and didn't commit it.

TJ-91 reviewed (commented):
>@wussler @TJ-91 Please check the naming of the key share intermediate values. There was a label "SHA3-256" that I couldn't make any sense of and which I removed. I also added "input to multiKeyCombine" as I assume this is was is meant here. There are still "(??)" marks in the text for this reason which we should remove before merging when my changes have been verified by you.

You are correct. Your suggestion to write `multiKeyCombine` makes more sense, since it's more high-level. 

You also replaced SHA3 here: 

> hex-encoded KMAC256 output is `TODO`.

I think it is more consistent to then also write `multiKeyCombine`.

falko-strenzke commented:
> You also replaced SHA3 here:
> 
> > hex-encoded KMAC256 output is `TODO`.
> 
> I think it is more consistent to then also write `multiKeyCombine`.

Changed it now also to `multiKeyCombine`.

falko-strenzke commented on draft-ietf-openpgp-pqc.md:
```diff
@@ -572,27 +555,22 @@ The change is planned to be integrated prior to IETF 121.\]
+                  || ecdhPublicKey || mlkemPublicKey || algId, oBits, domSep)
     return KEK
 
+Here, the parameters to KMAC256 appear in the order as specified in [SP800-186], Section 4, i.e., the key K, main input data X, requested output length L, and optional customization string S in that order.
```
done in bdce5bbb7c3f9e3cef400ab293dcb38dcd09fb7c

falko-strenzke commented on draft-ietf-openpgp-pqc.md:
```diff
@@ -615,21 +593,19 @@ The procedure to perform public-key encryption with an ML-KEM + ECDH composite s
  6. Compute `(mlkemCipherText, mlkemKeyShare) := ML-KEM.Encaps(mlkemPublicKey)`
 
- 7. Compute `fixedInfo` as specified in {{kem-fixed-info}}
+ 7. Compute `KEK := multiKeyCombine(ecdhKeyShare, ecdhCipherText, ecdhPublicKey, mlkemKeyShare, mlkemCipherText, mlkemPublicKey, algId, oBits=256)` as defined in {{kem-key-combiner}}
```
Removed `oBits` in e8ec2815000df0e7cdac12374d092e9b9aec60a5

fluppe2 commented on draft-ietf-openpgp-pqc.md:
```diff
@@ -951,16 +952,16 @@ The additional inclusion of `ecdhPublicKey` follows the security advice in Secti
 Hence, to be able to distinguish a key `K`, derived from shared keys `K1` and `K2` (with ciphertexts `C1` and `C2` and public keys `P1` and `P2`) as
 
-    K = SHA3-256(counter || K1 || C1 || P1 || K2 || C2 || P2 || fixedInfo)
+    K = KMAC256(K1 || K2 || C1 || C2 || P1 || P2 || algId, 256, domSep)
```
```suggestion
    K = KMAC256(K1 || K2, C1 || C2 || P1 || P2 || algId, 256, domSep)
```

wussler commented on draft-ietf-openpgp-pqc.md:
```diff
@@ -615,21 +620,19 @@ The procedure to perform public-key encryption with an ML-KEM + ECDH composite s
  1. Take the matching PKESK and own secret key packet as input
 
- 2. From the PKESK extract the algorithm ID and the `encryptedKey`, i.e., the wrapped session key
+ 2. From the PKESK extract the algorithm ID as `algId` and the the wrapped session key as `encryptedKey`
```
```suggestion
 2. From the PKESK extract the algorithm ID as `algId` and the wrapped session key as `encryptedKey`
```
